server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java

```diff
@@ -0,0 +1,264 @@
+        "Requires one of the following permissions:" +
+        "<ul>" +
+        "<li>'Administer System'</li>" +
+        "<li>'Administer' rights on the specified project</li>" +
+        "<li>'Browse' and 'See Source Code' on the specified project</li>" +
+        "</ul>",
+        PARAM_BASE_COMPONENT_ID, PARAM_BASE_COMPONENT_KEY))
+      .setSince("5.3")
+      .setResponseExample(getClass().getResource("tree-example.json"))
+      .setHandler(this)
+      .addSearchQuery("sonar", "component names", "component keys")
+      .addSortParams(newHashSet(SORTS), NAME_SORT, true)
+      .addPagingParams(100, MAX_SIZE);
+
+    action.createParam(PARAM_BASE_COMPONENT_ID)
+      .setDescription("base component id. The search is based on this component. It is not included in the response.")
+      .setExampleValue(UUID_EXAMPLE_02);
+
+    action.createParam(PARAM_BASE_COMPONENT_KEY)
+      .setDescription("base component key.The search is based on this component. It is not included in the response.")
+      .setExampleValue("org.apache.hbas:hbase");
+
+    action.createParam(PARAM_STRATEGY)
+      .setDescription("Strategy to search for base component children:" +
+        "<ul>" +
+        "<li>children: return the direct children components of the base component. Grandchildren components are not returned</li>" +
+        "<li>all: return all the children components of the base component. Grandchildren are returned.</li>" +
+        "<li>leaves: return all the children components (files, in general) which don't have other children. They are the leaves of the component tree.</li>" +
+        "</ul>")
+      .setPossibleValues(STRATEGIES)
+      .setDefaultValue(ALL_STRATEGY);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    TreeWsResponse treeWsResponse = doHandle(toTreeWsRequest(request));
+    writeProtobuf(treeWsResponse, request, response);
+  }
+
+  private TreeWsResponse doHandle(TreeWsRequest treeWsRequest) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto baseComponent = componentFinder.getByUuidOrKey(dbSession, treeWsRequest.getBaseComponentId(), treeWsRequest.getBaseComponentKey());
+      checkPermissions(baseComponent);
+      SnapshotDto baseSnapshot = dbClient.snapshotDao().selectLastSnapshotByComponentId(dbSession, baseComponent.getId());
+      if (baseSnapshot == null) {
+        return emptyResponse(treeWsRequest);
+      }
+
+      ComponentTreeQuery query = toComponentTreeQuery(treeWsRequest, baseSnapshot);
+      List<ComponentDto> components;
+      int total;
+      switch (treeWsRequest.getStrategy()) {
+        case CHILDREN_STRATEGY:
+          components = dbClient.componentDao().selectDirectChildren(dbSession, query);
+          total = dbClient.componentDao().countDirectChildren(dbSession, query);
+          break;
+        case LEAVES_STRATEGY:
+        case ALL_STRATEGY:
+        default:
```

Comment:
`default` should throw an IllegalStateException to avoid bad results the day a new constant value is added.